### PR TITLE
[dyld] Fix memory leaks when resolving symbols.

### PR DESCRIPTION
**Pull request by vgvassilev**
This was seen in an experimental branch of cmssw.

### CODE REVIEW

**Comment by vgvassilev** [comment]
cc: @davidlange6

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp [hunk 1]
```diff
@@ -869,6 +865,11 @@ namespace cling {
     return ""; // Search found no match.
   }
 
+  DynamicLibraryManager::~DynamicLibraryManager() {
+    static_assert(sizeof(Dyld) > 0, "Incomplete type");
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:869]: The next line is ill-formed without definition of `m-Dyld`'s type. What does this `static_assert`add?

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:869]: The static asserts enforces the compiler to know the size. Even if the next line was ill-formed, it compiled with incomplete type without even a warning. That caused the destructor for Dyld to not be called, leading to significant memory leaks.

Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:869]: Ill-formed means it doesn't compile. And I'm wrong: https://www.godbolt.org/z/1ovE34

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h [hunk 2]
```diff
@@ -97,6 +97,9 @@ namespace cling {
   public:
     DynamicLibraryManager(const InvocationOptions& Opts);
     ~DynamicLibraryManager();
+    DynamicLibraryManager(const DynamicLibraryManager&) = delete;
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:100]: Can we now make `m_Dyld` a `unique_ptr`, please?

**Comment by Axel-Naumann** [comment]
FYI, mac1015/cxx17 failures should be fixed.

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h (hunk 2 above)
Comment (vgvassilev) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:100]: No, we can't. The default constructor needs to be in the same file which is inconvenient (requires quite a few more includes).

Comment (Axel-Naumann) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:100]: I see no default ctor declared, and the compiler won't generate one as others are declared. There are several ctors but those are all outlined.

Comment (vgvassilev) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:100]: I meant the outlined ctor we have needs to be moved here.

Comment (Axel-Naumann) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:100]: O right. Yes, that sounds like an improvement!

Comment (Axel-Naumann) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:100]: O but you were saying it's inconvenient, right, unfortunate :-( Fair enough, I will keep forgetting why we cannot use unique_ptr until you reactor this ;-) Certainly good enough for now, thanks for explaining!

**Review by Axel-Naumann — APPROVED** [review]

**Pull request merged**